src/main/java/org/opensearch/security/privileges/PrivilegesEvaluatorImpl.java

```diff
@@ -0,0 +1,758 @@
+        this.dnfofForEmptyResultsEnabled = isDnfofEmptyEnabled(generalConfiguration);
+        this.filteredAliasMode = getFilteredAliasMode(generalConfiguration);
+
+        try {
+            RoleBasedActionPrivileges actionPrivileges = new RoleBasedActionPrivileges(rolesConfiguration, flattenedActionGroups, settings);
+            Metadata metadata = clusterStateSupplier.get().metadata();
+            actionPrivileges.updateStatefulIndexPrivileges(metadata.getIndicesLookup(), metadata.version());
+            RoleBasedActionPrivileges oldInstance = this.actionPrivileges.getAndSet(actionPrivileges);
+
+            if (oldInstance != null) {
+                oldInstance.clusterStateMetadataDependentPrivileges().shutdown();
+            }
+        } catch (Exception e) {
+            log.error("Error while updating ActionPrivileges", e);
+        }
+
+    }
+
+    @Override
+    public void updateClusterStateMetadata(ClusterService clusterService) {
+        RoleBasedActionPrivileges actionPrivileges = this.actionPrivileges.get();
+        if (actionPrivileges != null) {
+            actionPrivileges.clusterStateMetadataDependentPrivileges().updateClusterStateMetadataAsync(clusterService, threadPool);
+        }
+    }
+
+    @Override
+    public void shutdown() {
+        RoleBasedActionPrivileges roleBasedActionPrivileges = this.actionPrivileges.get();
+        if (roleBasedActionPrivileges != null) {
+            roleBasedActionPrivileges.clusterStateMetadataDependentPrivileges().shutdown();
+        }
+    }
+
+    @Override
+    public PrivilegesEvaluationContext createContext(User user, String action) {
+        return createContext(user, action, null, null, null);
+    }
+
+    @Override
+    public PrivilegesEvaluationContext createContext(
+        User user,
+        String action0,
+        ActionRequest request,
+        ActionRequestMetadata<?, ?> actionRequestMetadata,
+        Task task
+    ) {
+        TransportAddress caller = threadContext.getTransient(ConfigConstants.OPENDISTRO_SECURITY_REMOTE_ADDRESS);
+
+        ActionPrivileges actionPrivileges;
+        ImmutableSet<String> mappedRoles;
+
+        if (user.isPluginUser()) {
+            mappedRoles = ImmutableSet.of();
+            actionPrivileges = this.pluginIdToActionPrivileges.get(user.getName());
+            if (actionPrivileges == null) {
+                actionPrivileges = ActionPrivileges.EMPTY;
+            }
+        } else {
+            mappedRoles = this.roleMapper.map(user, caller);
+            actionPrivileges = this.actionPrivileges.get();
+        }
+
+        return new PrivilegesEvaluationContext(
+            user,
+            mappedRoles,
+            action0,
+            request,
+            task,
+            irr,
+            resolver,
+            clusterStateSupplier,
+            actionPrivileges
+        );
+    }
+
+    @Override
+    public PrivilegesEvaluatorResponse evaluate(PrivilegesEvaluationContext context) {
```

Comment:
as a follow-up, should this be broken down into Cluster/Index/Doc/Field-level evaluate methods?

Reply:
PrivilegeEvaluator is only for actions, so we could only break it into cluster vs index. 

Still, I am not sure of the use-case; could you elaborate on that?